File: sdk/src/main/java/ru/usedesk/sdk/internal/data/framework/api/standard/SocketApi.java
package ru.usedesk.sdk.internal.data.framework.api.standard;

import android.support.annotation.NonNull;

import com.google.gson.Gson;
import com.google.gson.JsonParseException;

import org.json.JSONException;
import org.json.JSONObject;

import java.net.HttpURLConnection;
import java.net.URISyntaxException;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

import javax.inject.Inject;

import io.socket.client.IO;
import io.socket.client.Socket;
import io.socket.emitter.Emitter;
import ru.usedesk.sdk.R;
import ru.usedesk.sdk.external.entity.chat.Constants;
import ru.usedesk.sdk.external.entity.chat.OnMessageListener;
import ru.usedesk.sdk.external.entity.chat.UsedeskActionListener;
import ru.usedesk.sdk.external.entity.exceptions.ApiException;
import ru.usedesk.sdk.internal.data.framework.api.standard.entity.request.BaseRequest;
import ru.usedesk.sdk.internal.data.framework.api.standard.entity.response.BaseResponse;
import ru.usedesk.sdk.internal.data.framework.api.standard.entity.response.ErrorResponse;
import ru.usedesk.sdk.internal.data.framework.api.standard.entity.response.InitChatResponse;
import ru.usedesk.sdk.internal.data.framework.api.standard.entity.response.NewMessageResponse;
import ru.usedesk.sdk.internal.data.framework.api.standard.entity.response.SendFeedbackResponse;
import ru.usedesk.sdk.internal.data.framework.api.standard.entity.response.SetEmailResponse;
import ru.usedesk.sdk.internal.utils.LogUtils;

public class SocketApi {
    private static final String TAG = SocketApi.class.getSimpleName();

    private final Map<String, Emitter.Listener> emitterListeners = new HashMap<>(5);

    private final Gson gson;

    private Socket socket;

    private UsedeskActionListener actionListener;
    private final Emitter.Listener disconnectEmitterListener = args -> {
        LogUtils.LOGE(TAG, "Disconnected.");
        actionListener.onDisconnected();
    };
    private final Emitter.Listener connectErrorEmitterListener = args -> {
        LogUtils.LOGE(TAG, "Error connecting: + " + Arrays.toString(args));
        actionListener.onError(R.string.message_connecting_error);
    };
    private OnMessageListener onMessageListener;
    private final Emitter.Listener connectEmitterListener = args -> {
        LogUtils.LOGD(TAG, "ConnectEmitterListener.args = " + Arrays.toString(args));
        onMessageListener.onInitChat();
    };
    private final Emitter.Listener baseEventEmitterListener = args -> {
        String rawResponse = args[0].toString();

        LogUtils.LOGD(TAG, "BaseEventEmitterListener.rawResponse = " + rawResponse);

        BaseResponse response = process(rawResponse);

        if (response != null) {
            switch (response.getType()) {
                case ErrorResponse.TYPE:
                    ErrorResponse errorResponse = (ErrorResponse) response;
                    if (HttpURLConnection.HTTP_FORBIDDEN == errorResponse.getCode()) {
                        onMessageListener.onTokenError();
                    }
                    break;
                case InitChatResponse.TYPE:
                    InitChatResponse initChatResponse = (InitChatResponse) response;
                    onMessageListener.onInit(initChatResponse.getToken(),
                            initChatResponse.getSetup());
                    break;
                case SetEmailResponse.TYPE:
                    break;
                case NewMessageResponse.TYPE:
                    NewMessageResponse newMessageResponse = (NewMessageResponse) response;
                    onMessageListener.onNew(newMessageResponse.getMessage());
                    break;
                case SendFeedbackResponse.TYPE:
                    onMessageListener.onFeedback();
                    break;
            }
        }
    };

    @Inject
    SocketApi(Gson gson) {
        this.gson = gson;
    }

    public boolean isConnected() {
        return socket.connected();
    }

    public void setSocket(@NonNull String url) throws ApiException {
        try {
            socket = IO.socket(url);
        } catch (URISyntaxException e) {
            throw new ApiException(e.getMessage());
        }
    }

    public void connect(UsedeskActionListener actionListener, OnMessageListener onMessageListener) {
        if (socket == null) {
            return;
        }

        this.actionListener = actionListener;
        this.onMessageListener = onMessageListener;

        emitterListeners.put(Constants.EVENT_SERVER_ACTION, baseEventEmitterListener);
        emitterListeners.put(Socket.EVENT_CONNECT_ERROR, connectErrorEmitterListener);
        emitterListeners.put(Socket.EVENT_CONNECT_TIMEOUT, connectErrorEmitterListener);
        emitterListeners.put(Socket.EVENT_DISCONNECT, disconnectEmitterListener);
        emitterListeners.put(Socket.EVENT_CONNECT, connectEmitterListener);

        for (String event : emitterListeners.keySet()) {
            socket.on(event, emitterListeners.get(event));
        }

        socket.connect();
    }

    public void disconnect() {
        for (String event : emitterListeners.keySet()) {
            socket.off(event, emitterListeners.get(event));
        }

        emitterListeners.clear();

        socket.disconnect();
    }

    public void emitterAction(BaseRequest baseRequest) throws ApiException {
        if (socket == null) {
            return;
        }

        try {
            JSONObject jsonObject = new JSONObject(gson.toJson(baseRequest));
            LogUtils.LOGD(TAG, "emitAction(). request = " + jsonObject);

            socket.emit(Constants.EVENT_SERVER_ACTION, jsonObject);
        } catch (JSONException e) {
            throw new ApiException(e.getMessage());
        }
    }

    private BaseResponse process(String rawResponse) {
        try {
            BaseRequest baseRequest = gson.fromJson(rawResponse, BaseRequest.class);

            if (baseRequest != null && baseRequest.getType() != null) {
                switch (baseRequest.getType()) {
                    case InitChatResponse.TYPE:
                        return gson.fromJson(rawResponse, InitChatResponse.class);
                    case ErrorResponse.TYPE:
                        return gson.fromJson(rawResponse, ErrorResponse.class);
                    case NewMessageResponse.TYPE:
                        return gson.fromJson(rawResponse, NewMessageResponse.class);
                    case SendFeedbackResponse.TYPE:
                        return gson.fromJson(rawResponse, SendFeedbackResponse.class);
                    case SetEmailResponse.TYPE:
                        return gson.fromJson(rawResponse, SetEmailResponse.class);
                }
            }
        } catch (JsonParseException e) {
            LogUtils.LOGE(TAG, e);
        }

        return null;
    }
}

File: chat-sdk/src/main/java/ru/usedesk/chat_sdk/internal/data/repository/api/loader/SocketApi.kt
package ru.usedesk.chat_sdk.internal.data.repository.api.loader

import com.google.gson.Gson
import com.google.gson.JsonParseException
import io.socket.client.IO
import io.socket.client.Socket
import io.socket.emitter.Emitter
import org.json.JSONException
import org.json.JSONObject
import ru.usedesk.chat_sdk.internal.data.repository.api._entity.request.BaseRequest
import ru.usedesk.chat_sdk.internal.data.repository.api._entity.request.InitChatRequest
import ru.usedesk.chat_sdk.internal.data.repository.api._entity.response.*
import ru.usedesk.chat_sdk.internal.data.repository.api._entity.response.ChatInitedResponse
import ru.usedesk.common_sdk.external.entity.exceptions.UsedeskException
import ru.usedesk.common_sdk.external.entity.exceptions.UsedeskSocketException
import toothpick.InjectConstructor
import java.net.HttpURLConnection
import java.net.URISyntaxException

@InjectConstructor
internal class SocketApi(
        private val gson: Gson
) {

    private val emitterListeners: MutableMap<String, Emitter.Listener> = hashMapOf()
    private var socket: Socket? = null

    private lateinit var eventListener: EventListener
    private lateinit var initChatRequest: InitChatRequest

    private val disconnectEmitterListener = Emitter.Listener {
        eventListener.onDisconnected()
    }

    private val connectErrorEmitterListener = Emitter.Listener {
        onConnectError()
    }

    private val connectEmitterListener = Emitter.Listener {
        eventListener.onConnected()
        sendRequest(initChatRequest)
    }

    private val baseEventEmitterListener = Emitter.Listener {
        onResponse(it[0].toString())
    }

    private fun onConnectError() {
        eventListener.onException(UsedeskSocketException(UsedeskSocketException.Error.DISCONNECTED))
    }

    private fun onResponse(rawResponse: String) {
        val response = process(rawResponse)
        if (response != null) {
            try {
                when (response.type) {
                    ErrorResponse.TYPE -> {
                        val errorResponse = response as ErrorResponse
                        val usedeskSocketException: UsedeskSocketException
                        usedeskSocketException = when (errorResponse.code) {
                            HttpURLConnection.HTTP_FORBIDDEN -> {
                                eventListener.onTokenError()
                                UsedeskSocketException(UsedeskSocketException.Error.FORBIDDEN_ERROR, errorResponse.message)
                            }
                            HttpURLConnection.HTTP_BAD_REQUEST -> UsedeskSocketException(UsedeskSocketException.Error.BAD_REQUEST_ERROR, errorResponse.message)
                            HttpURLConnection.HTTP_INTERNAL_ERROR -> UsedeskSocketException(UsedeskSocketException.Error.INTERNAL_SERVER_ERROR, errorResponse.message)
                            else -> UsedeskSocketException(UsedeskSocketException.Error.UNKNOWN_FROM_SERVER_ERROR, errorResponse.message)
                        }
                        eventListener.onException(usedeskSocketException)
                    }
                    ChatInitedResponse.TYPE -> {
                        eventListener.onInited(response as ChatInitedResponse)
                    }
                    EmailResponse.TYPE -> {
                    }
                    MessageResponse.TYPE -> {
                        eventListener.onNew(response as MessageResponse)
                    }
                    FeedbackResponse.TYPE -> eventListener.onFeedback()
                }
            } catch (e: Exception) {
                eventListener.onException(e)
            }
        }
    }

    fun isConnected(): Boolean {
        return socket?.connected() == true
    }

    @Throws(UsedeskException::class)
    fun connect(url: String,
                token: String?,
                companyId: String,
                eventListener: EventListener
    ) {
        if (socket != null) {
            return
        }
        socket = try {
            IO.socket(url)
        } catch (e: URISyntaxException) {
            throw UsedeskSocketException(UsedeskSocketException.Error.IO_ERROR, e.message)
        }

        this.eventListener = eventListener
        this.initChatRequest = InitChatRequest(token, companyId, url)

        emitterListeners[EVENT_SERVER_ACTION] = baseEventEmitterListener
        emitterListeners[Socket.EVENT_CONNECT_ERROR] = connectErrorEmitterListener
        emitterListeners[Socket.EVENT_CONNECT_TIMEOUT] = connectErrorEmitterListener
        emitterListeners[Socket.EVENT_DISCONNECT] = disconnectEmitterListener
        emitterListeners[Socket.EVENT_CONNECT] = connectEmitterListener

        for (event in emitterListeners.keys) {
            socket?.on(event, emitterListeners[event])
        }
        socket?.connect()
    }

    fun disconnect() {
        for (event in emitterListeners.keys) {
            socket?.off(event, emitterListeners[event])
        }
        emitterListeners.clear()
        socket?.disconnect()
    }

    @Throws(UsedeskSocketException::class)
    fun sendRequest(baseRequest: BaseRequest) {
        try {
            val jsonObject = JSONObject(gson.toJson(baseRequest))
            socket?.emit(EVENT_SERVER_ACTION, jsonObject)
        } catch (e: JSONException) {
            throw UsedeskSocketException(UsedeskSocketException.Error.JSON_ERROR, e.message)
        }
    }

    private fun process(rawResponse: String): BaseResponse? {
        try {
            val baseRequest = gson.fromJson(rawResponse, BaseRequest::class.java)
            when (baseRequest.type) {
                ChatInitedResponse.TYPE -> ChatInitedResponse::class.java
                ErrorResponse.TYPE -> ErrorResponse::class.java
                MessageResponse.TYPE -> MessageResponse::class.java
                FeedbackResponse.TYPE -> FeedbackResponse::class.java
                EmailResponse.TYPE -> EmailResponse::class.java
                else -> null
            }?.also {
                return gson.fromJson(rawResponse, it)
            }
        } catch (e: JsonParseException) {
            eventListener.onException(UsedeskSocketException(UsedeskSocketException.Error.JSON_ERROR, e.message))
        }
        return null
    }

    companion object {
        private const val EVENT_SERVER_ACTION = "dispatch"
    }

    interface EventListener {
        fun onConnected()
        fun onDisconnected()
        fun onTokenError()
        fun onFeedback()
        fun onException(exception: Exception)

        fun onInited(chatInitedResponse: ChatInitedResponse)
        fun onNew(messageResponse: MessageResponse)
    }
}